cmd/basin/publication.go

```diff
@@ -211,7 +229,41 @@ func newPublicationStartCommand() *cli.Command {
 			}
 			defer bp.Close()
 
-			basinStreamer := app.NewBasinStreamer(ns, r, bp, privateKey)
+			pgxConn, err := pgx.Connect(cCtx.Context, connString)
+			if err != nil {
+				return fmt.Errorf("connect: %s", err)
+			}
+			defer func() {
+				_ = pgxConn.Close(cCtx.Context)
+			}()
+
+			tx, err := pgxConn.Begin(cCtx.Context)
+			if err != nil {
+				return fmt.Errorf("failed to begin transaction")
+			}
+			defer func() {
+				if err != nil {
+					_ = tx.Rollback(cCtx.Context)
+				}
+			}()
+
+			cols, err := inspectTable(cCtx.Context, tx, rel)
+			if err != nil {
+				return fmt.Errorf("failed to inspect source table: %s", err)
+			}
+
+			// Creates a new db manager when replication starts
+			dbDir := path.Join(dir, publication)
+			winSize := time.Duration(cfg.Publications[publication].WindowSize) * time.Second
+			uploader := app.NewBasinUploader(ns, rel, bp, privateKey)
+			dbm := app.NewDBManager(dbDir, rel, cols, winSize, uploader)
+
+			// Before starting replication, upload the remaining data
+			if err := dbm.UploadAll(cCtx.Context); err != nil {
+				return fmt.Errorf("upload all: %s", err)
+			}
+
+			basinStreamer := app.NewBasinStreamer(ns, r, dbm)
```

Comment:
Streamer doesn't bind to `BasinProvider` anymore because going forward, provider is only used for metadata. `Uploader` is used to `upload` files.  `DBManager` is instantiated with an `Uploader`.

Reply:
Just a comment about an alternative more decoupled architecture. No need to change it. 

One option would be to not have the `Uploader` as a dependency of `DBManager`. You could use a channel to synchronize both components and make them independent. Every time a file is ready to be uploaded, `DBManager` notifies the channel, `Uploader` picks up the notification and uploads the file. Notice that I'm not saying to make it async, it would still be blocking, but more decoupled and easier to test the parts independently.